src/client/serialize/daos_serialize.c

```diff
@@ -0,0 +1,950 @@
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[3];
+	rc = serialize_uint(file_id, entry, "DAOS_PROP_CO_CSUM");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[4];
+	rc = serialize_uint(file_id, entry,
+			    "DAOS_PROP_CO_CSUM_CHUNK_SIZE");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[5];
+	rc = serialize_uint(file_id, entry,
+			    "DAOS_PROP_CO_CSUM_SERVER_VERIFY");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[6];
+	rc = serialize_uint(file_id, entry, "DAOS_PROP_CO_REDUN_FAC");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[7];
+	rc = serialize_uint(file_id, entry, "DAOS_PROP_CO_REDUN_LVL");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[8];
+	rc = serialize_uint(file_id, entry, "DAOS_PROP_CO_SNAPSHOT_MAX");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[9];
+	rc = serialize_uint(file_id, entry, "DAOS_PROP_CO_COMPRESS");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[10];
+	rc = serialize_uint(file_id, entry, "DAOS_PROP_CO_ENCRYPT");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[11];
+	rc = serialize_str(file_id, entry, "DAOS_PROP_CO_OWNER");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[12];
+	rc = serialize_str(file_id, entry, "DAOS_PROP_CO_OWNER_GROUP");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[13];
+	rc = serialize_uint(file_id, entry, "DAOS_PROP_CO_DEDUP");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[14];
+	rc = serialize_uint(file_id, entry,
+				      "DAOS_PROP_CO_DEDUP_THRESHOLD");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	entry = &prop_query->dpp_entries[15];
+	rc = serialize_uint(file_id, entry, "DAOS_PROP_CO_ALLOCED_OID");
+	if (rc != 0) {
+		D_GOTO(out, rc);
+	}
+
+	/* serialize ACL */
+	if (prop_query->dpp_nr > 16) {
+		entry = &prop_query->dpp_entries[16];
+		rc = serialize_acl(file_id, entry, "DAOS_PROP_CO_ACL");
+		if (rc != 0) {
+			D_GOTO(out, rc);
+		}
+	}
+out:
+	return rc;
+}
+
+int
+daos_cont_serialize_attrs(hid_t file_id, hid_t *usr_attr_memtype,
+			  int num_attrs, char **names, char **buffers,
+			  size_t *sizes)
+{
+	int		rc = 0;
+	hid_t		status = 0;
+	hid_t		dset = 0;
+	hid_t		dspace = 0;
+	hsize_t		dims[1];
+	struct usr_attr	*attr_data = NULL;
+	int		i;
+
+	if (num_attrs == 0) {
+		D_GOTO(out_no_attrs, rc);
+	}
+
+	if (names == NULL || buffers == NULL || sizes == NULL) {
+		rc = -DER_NOMEM;
+		D_GOTO(out, rc);
+	}
+
+	/* Create the user attribute data space */
+	dims[0] = num_attrs;
+	dspace = H5Screate_simple(1, dims, NULL);
+	if (dspace < 0) {
+		rc = -DER_MISC;
+		D_GOTO(out, rc);
+	}
+
+	/* Create the user attribute dataset */
+	dset = H5Dcreate(file_id, SERIALIZE_ATTR_DSET, *usr_attr_memtype,
```

Comment:
maybe we discussed this before and we agreed that what you had was better, but wouldn't it be more consistent to use HDF5 attributes like you do for properties, instead of a dataset for user attributes?

Reply:
I guess maybe a dataset is better because we don't know how many attributes we have and what they are..

Reply:
@mchaarawi yeah that is ultimately why it was done this way.